cmd/sync-rover-groups/main.go

```diff
@@ -0,0 +1,164 @@
+	fs.StringVar(&opts.groupsFile, "groups-file", "/tmp/groups.yaml", "The file to store the groups in yaml format")
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatal("could not parse args")
+	}
+	return opts
+}
+
+func (o *options) validate() error {
+	level, err := logrus.ParseLevel(o.logLevel)
+	if err != nil {
+		return fmt.Errorf("invalid log level specified: %w", err)
+	}
+	logrus.SetLevel(level)
+
+	values := o.manifestDirRaw.Strings()
+	if len(values) == 0 {
+		return fmt.Errorf("--manifest-dir must be set")
+	}
+	o.manifestDirs = sets.NewString()
+	o.manifestDirs.Insert(values...)
+	return nil
+}
+
+func addSchemes() error {
+	if err := userv1.AddToScheme(scheme.Scheme); err != nil {
+		return fmt.Errorf("failed to add userv1 to scheme: %w", err)
+	}
+	if err := rbacv1.AddToScheme(scheme.Scheme); err != nil {
+		return fmt.Errorf("failed to add rbacv1 to scheme: %w", err)
+	}
+	if err := templatev1.AddToScheme(scheme.Scheme); err != nil {
+		return fmt.Errorf("failed to add templatev1 to scheme: %w", err)
+	}
+	return nil
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	opts := parseOptions()
+
+	if err := opts.validate(); err != nil {
+		logrus.WithError(err).Fatal("failed to validate the option")
+	}
+
+	if err := addSchemes(); err != nil {
+		logrus.WithError(err).Fatal("failed to add schemes")
+	}
+
+	conn, err := ldapv3.DialURL(fmt.Sprintf("ldap://%s", opts.ldapServer))
+	if err != nil {
+		logrus.Fatal(err)
+	}
+	defer conn.Close()
+
+	groupCollector := newYamlGroupCollector(opts.validateSubjects)
+	groupResolver := &ldapGroupResolver{conn: conn}
+	groups, err := roverGroups(opts.manifestDirs, groupCollector, groupResolver)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to get rover groups")
+	}
+	data, err := yaml.Marshal(groups)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to marshal groups")
+	}
+	if err := ioutil.WriteFile(opts.groupsFile, data, 0644); err != nil {
+		logrus.WithError(err).WithField("file", opts.groupsFile).Fatal("failed to write file")
+	}
+}
+
+type Group struct {
+	Name    string   `json:"name"`
+	Members []string `json:"members"`
+}
+
+type groupResolver interface {
+	resolve(name string) (*Group, error)
+}
+
+type groupCollector interface {
+	collect(dir string) (sets.String, error)
+}
+
+func roverGroups(manifestDirs sets.String, groupCollector groupCollector, groupResolver groupResolver) (map[string][]string, error) {
+	var errs []error
+
+	groupNames := sets.NewString()
+	for _, d := range manifestDirs.List() {
+		names, err := groupCollector.collect(d)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to collect groups for %s: %w", d, err))
+			continue
+		}
+		groupNames.Insert(names.List()...)
+	}
+
+	groupNames.Insert(api.CIAdminsGroupName)
+
+	groups := map[string][]string{}
+	for _, name := range groupNames.List() {
+		logrus.WithField("group", name).Debug("resolving group ...")
+		g, err := groupResolver.resolve(name)
```

Comment:
Could we reasonably parallelize? I imagine this could get slow on I/O if we have more groups.

Reply:
It is quick enough for the current amount of groups.
We could improve this later in the future if necessary.